src/jit/lowerxarch.cpp

```diff
@@ -805,6 +805,77 @@ void Lowering::LowerHWIntrinsicCC(GenTreeHWIntrinsic* node, NamedIntrinsic newIn
     }
 }
 
+//----------------------------------------------------------------------------------------------
+// LowerFusedMultiplyAdd: Changes NI_FMA_MultiplyAddScalar produced by Math(F).FusedMultiplyAdd
+//     to a better FMA intrinsics if there are GT_NEG around in order to eliminate them.
+//
+//  Arguments:
+//     node - The hardware intrinsic node
+//
+//  Notes:
+//     Math(F).FusedMultiplyAdd is expanded into NI_FMA_MultiplyAddScalar and
+//     depending on additional GT_NEG nodes around it can be:
+//
+//      x *  y + z -> NI_FMA_MultiplyAddScalar
+//      x * -y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x *  y + z -> NI_FMA_MultiplyAddNegatedScalar
+//     -x * -y + z -> NI_FMA_MultiplyAddScalar
+//      x *  y - z -> NI_FMA_MultiplySubtractScalar
+//      x * -y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x *  y - z -> NI_FMA_MultiplySubtractNegatedScalar
+//     -x * -y - z -> NI_FMA_MultiplySubtractScalar
+//
+void Lowering::LowerFusedMultiplyAdd(GenTreeHWIntrinsic* node)
+{
+    GenTreeArgList*     argList = node->gtGetOp1()->AsArgList();
+    GenTreeHWIntrinsic* createScalarOps[4]; // 3 arguments with 1-based index
+
+    for (int i = 1; i < 4; i++)
+    {
+        assert(argList->Current() != nullptr);
+        if (argList->Current()->OperIsHWIntrinsic())
+        {
+            GenTreeHWIntrinsic* hwArg = argList->Current()->AsHWIntrinsic();
+            if (hwArg->gtHWIntrinsicId == NI_Vector128_CreateScalarUnsafe)
+            {
+                createScalarOps[i] = hwArg;
+                argList            = argList->Rest();
+                continue;
+            }
+        }
+
+        return; // Math(F).FusedMultiplyAdd is expected to emit three NI_Vector128_CreateScalarUnsafe
```

Comment:
I'm not sure this logic is the easiest to follow.

It's not immediately obvious (at least it wasn't to me 😄) that the `for` loop exits on the good path due to `i == 4` and that this isn't hit in the "good" case due to the `continue`.